File: Day 15/Python/sol.py
"""
FILE: Day 15/Python/sol.py
Author: Alex Jones
Desc: Solution to day 15 problems (29 & 30) for Advent of Code 2024, solved in Python 3.
"""
data = open("Day 15/data.txt", "r").read().strip().replace("\r","").split("\n\n")
grid = [[c for c in row] for row in data[0].split("\n")]
moves = data[1].replace("\n","")

move_map = {"<": (0, -1), "^": (-1, 0), ">": (0, 1), "v": (1, 0)}
widen_map = {'@': "@.", 'O': "[]", '#': "##", '.': ".."}

def can_move(grid, pos, dir_) -> bool:
    (ny, nx) = (pos[0] + dir_[0], pos[1] + dir_[1])
    if grid[ny][nx] in '[]':
        left  = (ny, nx) if grid[ny][nx] == '[' else (ny, nx - 1)
        right = (ny, nx) if grid[ny][nx] == ']' else (ny, nx + 1)
        return (dir_ == (0,  1) or can_move(grid, left, dir_)) and \
               (dir_ == (0, -1) or can_move(grid, right, dir_))
    elif grid[ny][nx] == 'O':
        return can_move(grid, (ny, nx), dir_)
    return grid[ny][nx] != '#'

def make_move(grid, pos, dir_, partial=False):
    (ny, nx) = (pos[0] + dir_[0], pos[1] + dir_[1])
    if not partial and dir_[0] != 0:
        if grid[pos[0]][pos[1]] == '[':
            make_move(grid, (pos[0], pos[1] + 1), dir_, partial=True)
        elif grid[pos[0]][pos[1]] == ']':
            make_move(grid, (pos[0], pos[1] - 1), dir_, partial=True)
    if grid[ny][nx] != '.':
        make_move(grid, (ny, nx), dir_)
    grid[ny][nx] = grid[pos[0]][pos[1]]
    grid[pos[0]][pos[1]] = '.'

def simulate(grid, moves):
    robot = min(((i, j)) for i, r in enumerate(grid) for j, c in enumerate(r) if c == '@')
    for move in moves:
        dir_ = move_map[move]
        if can_move(grid, robot, dir_):
            make_move(grid, robot, dir_)
            robot = (robot[0] + dir_[0], robot[1] + dir_[1])

def get_total_gps(grid):
    return sum(100 * i + j for i, r in enumerate(grid) for j, c in enumerate(r) if c in 'O[')

p1 = [r.copy() for r in grid]
simulate(p1, moves)
print("Problem 29:", get_total_gps(p1))

p2 = [list("".join(widen_map[c] for c in r)) for r in grid]
simulate(p2, moves)
print("Problem 30:", get_total_gps(p2))

File: Day 15/Rust/sol.rs
/*
 * FILE: Day 15/Rust/sol.rs
 * Author: Alex Jones
 * Desc: Solution to day 15 problems (29 & 30) for Advent of Code 2024, solved in Rust.
 */
use itertools::Itertools;
use std::fs;

fn move_to_dir(move_char: u8) -> Option<(i32, i32)> {
    match move_char {
        b'<' => Some((0, -1)),
        b'^' => Some((-1, 0)),
        b'>' => Some((0, 1)),
        b'v' => Some((1, 0)),
        _ => None,
    }
}

fn widened(pos_char: u8) -> [u8; 2] {
    match pos_char {
        b'@' => [b'@', b'.'],
        b'O' => [b'[', b']'],
        c => [c, c],
    }
}

fn can_move(grid: &Vec<Vec<u8>>, pos: (i32, i32), dir: (i32, i32)) -> bool {
    let (ny, nx) = (pos.0 + dir.0, pos.1 + dir.1);
    let char = grid[ny as usize][nx as usize];
    if char == b'[' || char == b']' {
        let left = if char == b'[' { (ny, nx) } else { (ny, nx - 1) };
        let right = if char == b']' { (ny, nx) } else { (ny, nx + 1) };
        (dir == (0, 1) || can_move(grid, left, dir))
            && (dir == (0, -1) || can_move(grid, right, dir))
    } else if char == b'O' {
        can_move(grid, (ny, nx), dir)
    } else {
        char != b'#'
    }
}

fn make_move(grid: &mut Vec<Vec<u8>>, pos: (i32, i32), dir: (i32, i32), partial: bool) {
    let (ny, nx) = (pos.0 + dir.0, pos.1 + dir.1);
    let char = grid[pos.0 as usize][pos.1 as usize];
    if !partial && dir.0 != 0 {
        if char == b'[' {
            make_move(grid, (pos.0, pos.1 + 1), dir, true);
        } else if char == b']' {
            make_move(grid, (pos.0, pos.1 - 1), dir, true);
        }
    }
    if grid[ny as usize][nx as usize] != b'.' {
        make_move(grid, (ny, nx), dir, false);
    }
    grid[ny as usize][nx as usize] = char;
    grid[pos.0 as usize][pos.1 as usize] = b'.';
}

fn simulate(grid: &mut Vec<Vec<u8>>, moves: &str) {
    let mut robot = None;
    for (i, row) in grid.iter().enumerate() {
        for (j, &char) in row.iter().enumerate() {
            if char == b'@' {
                robot = Some((i as i32, j as i32));
                break;
            }
        }
    }
    let mut robot = robot.unwrap();
    for move_char in moves.bytes() {
        let dir = move_to_dir(move_char).unwrap();
        if can_move(grid, robot, dir) {
            make_move(grid, robot, dir, false);
            robot = (robot.0 + dir.0, robot.1 + dir.1);
        }
    }
}

fn get_total_gps(grid: &[Vec<u8>]) -> u32 {
    let mut total = 0u32;
    for (i, row) in grid.iter().enumerate() {
        for (j, &char) in row.iter().enumerate() {
            if char == b'O' || char == b'[' {
                total += 100 * (i as u32) + (j as u32);
            }
        }
    }
    total
}

pub fn solve() {
    let contents: String = fs::read_to_string("Day 15/data.txt")
        .expect("Could not read data file")
        .trim()
        .replace("\r", "");
    let data: (&str, &str) = contents.split("\n\n").collect_tuple().unwrap();

    let grid: Vec<Vec<u8>> = data.0.split("\n").map(|s| s.bytes().collect()).collect();
    let moves = data.1.replace('\n', "");

    let mut p1 = grid.clone();
    simulate(&mut p1, &moves);
    println!("Problem 29: {}", get_total_gps(&p1));

    let mut p2: Vec<Vec<u8>> = grid
        .iter()
        .map(|row| row.iter().flat_map(|&c| widened(c)).collect())
        .collect();
    simulate(&mut p2, &moves);
    println!("Problem 30: {}", get_total_gps(&p2));
}
